libcudacxx/include/cuda/__iterator/offset_iterator.h

```diff
@@ -0,0 +1,529 @@
+  _LIBCUDACXX_HIDE_FROM_ABI constexpr offset_iterator operator++(int)
+  {
+    offset_iterator __tmp = *this;
+    ++__offset_;
+    return __tmp;
+  }
+
+  //! @brief Decrements the stored offset
+  _CCCL_EXEC_CHECK_DISABLE
+  _LIBCUDACXX_HIDE_FROM_ABI constexpr offset_iterator& operator--()
+  {
+    --__offset_;
+    return *this;
+  }
+
+  //! @brief Decrements the stored offset
+  _CCCL_EXEC_CHECK_DISABLE
+  _LIBCUDACXX_HIDE_FROM_ABI constexpr offset_iterator operator--(int)
+  {
+    offset_iterator __tmp = *this;
+    --__offset_;
+    return __tmp;
+  }
+
+  //! @brief Advances the stored offset by \p __n
+  //! @param __n The number of elements to advance
+  _CCCL_EXEC_CHECK_DISABLE
+  [[nodiscard]] _LIBCUDACXX_HIDE_FROM_ABI constexpr offset_iterator operator+(difference_type __n) const
+  {
+    if constexpr (_CUDA_VSTD::__integer_like<_OffsetType>)
+    {
+      return offset_iterator{__iter_, __offset_ + static_cast<__offset_difference_t>(__n)};
+    }
+    else
+    {
+      return offset_iterator{__iter_, _CUDA_VSTD::next(__offset_, static_cast<__offset_difference_t>(__n))};
+    }
+  }
+
+  //! @brief Returns a copy of \p __x advanced by \p __n
+  //! @param __n The number of elements to advance
+  //! @param __x The original \c offset_iterator
+  _CCCL_EXEC_CHECK_DISABLE
+  [[nodiscard]] _LIBCUDACXX_HIDE_FROM_ABI friend constexpr offset_iterator
+  operator+(difference_type __n, const offset_iterator& __x)
+  {
+    return __x + __n;
+  }
+
+  //! @brief Advances the \c offset_iterator by \p __n
+  //! @param __n The number of elements to advance
+  _CCCL_EXEC_CHECK_DISABLE
+  _LIBCUDACXX_HIDE_FROM_ABI constexpr offset_iterator& operator+=(difference_type __n)
+  {
+    if constexpr (_CUDA_VSTD::__integer_like<_OffsetType>)
+    {
+      __offset_ += static_cast<__offset_difference_t>(__n);
+    }
+    else
+    {
+      _CUDA_VSTD::advance(__offset_, static_cast<__offset_difference_t>(__n));
+    }
+    return *this;
+  }
+
+  //! @brief Returns a copy of the \c offset_iterator decremented by \p __n
+  //! @param __n The number of elements to decrement
+  _CCCL_EXEC_CHECK_DISABLE
+  [[nodiscard]] _LIBCUDACXX_HIDE_FROM_ABI constexpr offset_iterator operator-(difference_type __n) const
+  {
+    if constexpr (_CUDA_VSTD::__integer_like<_OffsetType>)
+    {
+      return offset_iterator{__iter_, __offset_ - static_cast<__offset_difference_t>(__n)};
+    }
+    else
+    {
+      return offset_iterator{__iter_, _CUDA_VSTD::prev(__offset_, static_cast<__offset_difference_t>(__n))};
+    }
+  }
+
+  //! @brief Returns the distance between \p __lhs and \p __rhs
+  //! @param __lhs The left \c offset_iterator
+  //! @param __rhs The right \c offset_iterator
+  //! @return __rhs.offset() - __lhs.offset()
+  [[nodiscard]] _LIBCUDACXX_HIDE_FROM_ABI friend constexpr difference_type
+  operator-(const offset_iterator& __lhs, const offset_iterator& __rhs)
+  {
+    return static_cast<difference_type>(__rhs.offset() - __lhs.offset());
+  }
```

Comment:
I believe this is actually wrong. We need to return the distance of the memory location not just the offset.

This would result in 2 iterators comparing unequal even if their distance is non-zero and vice versa

@bernhardmgruber what do you think, that would be a slightly breaking change

Reply:
That said, it is what other iterators from ranges return so we might want to stay consistent

Reply:
`thrust::offset_iterator` does take the base iterator into account.